lib/prometheus/middleware/collector.rb

```diff
@@ -0,0 +1,79 @@
+# encoding: UTF-8
+
+require 'prometheus/client'
+
+module Prometheus
+  module Middleware
+    # Collector is a Rack middleware that provides a sample implementation of
+    # a HTTP tracer. The default label builder can be modified to export a
+    # different set of labels per recorded metric.
+    class Collector
+      attr_reader :app, :registry
+
+      def initialize(app, options = {}, &label_builder)
+        @app = app
+        @registry = options[:registry] || Client.registry
+        @label_builder = label_builder || DEFAULT_LABEL_BUILDER
+
+        init_request_metrics
+        init_exception_metrics
+      end
+
+      def call(env) # :nodoc:
+        trace(env) { @app.call(env) }
+      end
+
+      protected
+
+      DEFAULT_LABEL_BUILDER = proc do |env|
+        {
+          method: env['REQUEST_METHOD'].downcase,
+          path:   env['PATH_INFO'].to_s,
+        }
+      end
+
+      def init_request_metrics
+        @requests = @registry.counter(
+          :http_server_requests_total,
+          'The total number of HTTP requests handled by the Rack application.',
```

Comment:
Rack is the de facto standard interface for ruby HTTP servers. It can be implemented by many different servers (unicorn, thin, puma, ...). I thought about including it, but concluded after some discussions that including it in the metric name is neither expected nor helpful.